sdk/identity/Azure.Identity/src/MsalCacheReader.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.Identity.Client;
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using System.Security.Cryptography;
+
+namespace Azure.Identity
+{
+    internal class MsalCacheReader
+    {
+        private string _cachePath;
+        private string _cacheLockPath;
+        private int _cacheRetryCount;
+        private TimeSpan _cacheRetryDelay;
+        private DateTimeOffset _lastReadTime;
+
+        public MsalCacheReader(ITokenCache cache, string cachePath, int cacheRetryCount, TimeSpan cacheRetryDelay)
+        {
+            _cachePath = cachePath;
+
+            _cacheLockPath = cachePath + ".lock";
+
+            _cacheRetryCount = cacheRetryCount;
+
+            _cacheRetryDelay = cacheRetryDelay;
+
+            cache.SetBeforeAccessAsync(OnBeforeAccessAsync);
+        } 
+
+        private async Task OnBeforeAccessAsync(TokenCacheNotificationArgs args)
+        {
+            try
+            {
+                if (File.Exists(_cachePath) && _lastReadTime < File.GetLastWriteTimeUtc(_cachePath))
+                {
+                    using (var cacheLock = await SentinalFileLock.AquireAsync(_cacheLockPath, _cacheRetryCount, _cacheRetryDelay).ConfigureAwait(false))
+                    {
+                        byte[] cacheBytesFromDisk = await ReadCacheFromProtectedStorageAsync().ConfigureAwait(false);
+
+                        // update the last read time before deserialization so if deserialization fails we won't continually read the invalid file
+                        _lastReadTime = DateTimeOffset.UtcNow;
```

Comment:
If we are comparing to LastWriteTimeUtc in initial check shouldn't we setting _lastReadTime to `File.GetLastWriteTimeUtc`? Otherwise, we might miss a change that happened in between our check an setting `_lastReadTime = DateTimeOffset.UtcNow;`